XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed. This is the only callback that does not receive the context as
+// "this". Because it is called while the associated class is being destructed
+// it is unsafe to operate on the object.
+function destroy(id) { }
+
+// Add a set of hooks to be called during the lifetime of asynchronous events.
+// createHook() returns an instance of the AsyncHook constructor that can
+// control how those hooks are used. For example, enabling or disabling their
+// execution.
+asyncHook = async_wrap.createHook({ init, pre, post, destroy });
```

Comment:
`disable()` is a flag that toggles which set of callbacks need to be called for that instance. so if there's only one instance and it's disabled then the normal checks won't anything needs to be called.

since even after a hook has been disabled, it will continue to receive notifications on async paths already taken (100% intentional design). if you want an api that removes even those calls then that comes with the caveat that resource cleanup is then on the user I can think of a way to make that possible.

> Maybe better as `new Hook()`?

I could live with that.